sympy/assumptions/assume.py

```diff
@@ -86,80 +131,217 @@ def arg(self):
         x + 1
 
         """
-        return self._args[1]
+        args = self.args
+        if len(args) == 1:
+            return args[0]
+        return args
 
     @property
     def args(self):
-        return self._args[1:]
+        return self._args[1]
```

Comment:
Why is this method needed? Can we not just let the `Basic.args` property return the args as normal?

Reply:
Normal way will return `AppliedPredicate` for `Q.even(2).func` and `(Q.even, (2,))` for `Q.even(2).args`.

By overriding `args` like this, we can make `Q.even(2).func` return `Q.even` and `Q.even(2).args` return `(2,)`.  I think this way is better since we should give different heads for different predicates, such as `Q.even()`, `Q.odd()`, `Q.positive()`, ... etc.

Reply:
> I think this way is better since we should give different heads for different predicates

I don't understand what this means.

I think that almost all cases where args is overridden like this in the sympy codebase are warts to be removed. This always leads to problems down the line and it would generally be better if we could just accept the Basic model of func and args rather than fighting it.

(I realise that this wasn't added by you)

Reply:
Make sense. I will try fixing it.

Reply:
I tried to fix this, but `func` and `args` was used everywhere in assumptions module that it exceeds the scope of this PR. I will make another PR to fix this.

Reply:
Fair enough.

In general I think that if `args` is different from `_args` then something is wrong. At least they could be made equal and `.func` could be made to work with whatever is stored in `_args`.

Generally though everything is smoother if `obj.func` is `type(obj)` and `_args` is the args that would be passed to it. This eliminates the needs for a lot of hacky workarounds.

Having the predicate not in `args` means it can't be substituted e.g.:
```python
In [1]: p = Q.even(x)

In [2]: p
Out[2]: Q.even(x)

In [3]: p.subs(Q.even, Q.odd)
Out[3]: Q.even(x)
```
The analogy with functions is
```python
In [4]: sin(x).subs(sin, cos)
Out[4]: cos(x)

In [5]: f(x).subs(f, g)
Out[5]: g(x)
```
This is made to work in a hacky way because `sin`, `cos`, `f` and `g` are not `Basic`. If they were `Basic` and `f(x)` was just `Call(Function('f'), Symbol('x'))` then subs would work automatically. That's what what we want to aim for in Function and it's one of the chief advantages of having the function objects themselves be `Basic`. Disallowing that in the case of `Predicate` seems backward to me.

At the same time code should not make any presumptions about what `obj.func` is except that it can be called with the args as `obj.func(*obj.args)` so if `obj.func` is being used for anything else then that should be changed as well.